### PR TITLE
refactor(circuits): remove unnecessary inputs

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Prove reputation circuit accepts reputation nullifiers as input.

## What is the new behavior?

Prove reputation circuit doesn't accept reputation nullifiers as input.
Remove inputs that can be calculated.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by vimwitch** [comment]
Benchmark before:
```
Intel Core i5, 16 GB RAM

GST depth 10, UST depth 9: 2700 ms (2.7 s)
GST depth 12, UST depth 9: 2739 ms (2.739 s)
GST depth 14, UST depth 9: 2813 ms (2.813 s)
GST depth 16, UST depth 9: 2832 ms (2.832 s)
GST depth 18, UST depth 9: 2887 ms (2.887 s)
GST depth 20, UST depth 9: 2939 ms (2.939 s)
GST depth 22, UST depth 9: 2995 ms (2.995 s)
GST depth 24, UST depth 9: 3502 ms (3.502 s)
GST depth 26, UST depth 9: 3563 ms (3.563 s)
GST depth 28, UST depth 9: 3611 ms (3.611 s)
GST depth 30, UST depth 9: 3660 ms (3.66 s)
GST depth 32, UST depth 9: 3689 ms (3.689 s)
```

After:
```
Intel Core i5, 16 GB RAM

GST depth 10, UST depth 9: 2665 ms (2.665 s)
GST depth 12, UST depth 9: 2676 ms (2.676 s)
GST depth 14, UST depth 9: 2762 ms (2.762 s)
GST depth 16, UST depth 9: 2803 ms (2.803 s)
GST depth 18, UST depth 9: 2847 ms (2.847 s)
GST depth 20, UST depth 9: 2898 ms (2.898 s)
GST depth 22, UST depth 9: 2934 ms (2.934 s)
GST depth 24, UST depth 9: 2979 ms (2.979 s)
GST depth 26, UST depth 9: 3020 ms (3.02 s)
GST depth 28, UST depth 9: 3076 ms (3.076 s)
GST depth 30, UST depth 9: 3119 ms (3.119 s)
GST depth 32, UST depth 9: 3151 ms (3.151 s)
```

Roughly half second reduction in proving time for large depths.

**Comment by vimwitch** [comment]
Final benchmark with refactors:
```
Intel Core i5, 16 GB RAM

GST depth 10, UST depth 9: 2630 ms (2.63 s)
GST depth 12, UST depth 9: 2718 ms (2.718 s)
GST depth 14, UST depth 9: 2757 ms (2.757 s)
GST depth 16, UST depth 9: 2787 ms (2.787 s)
GST depth 18, UST depth 9: 2834 ms (2.834 s)
GST depth 20, UST depth 9: 2893 ms (2.893 s)
GST depth 22, UST depth 9: 2907 ms (2.907 s)
GST depth 24, UST depth 9: 2977 ms (2.977 s)
GST depth 26, UST depth 9: 2995 ms (2.995 s)
GST depth 28, UST depth 9: 3070 ms (3.07 s)
GST depth 30, UST depth 9: 3111 ms (3.111 s)
GST depth 32, UST depth 9: 3159 ms (3.159 s)
```

**Comment by vimwitch** [comment]
This is pretty much ready. Just need some final circuit tests to verify the output GST hashes and epoch keys.

**Comment by vimwitch** [comment]
- [ ] Reputation hash: Poseidon(4)
- [ ] Epk Nullifier: Poseidon(3) (need to transform the identity nullifier, could just add 1)
- [ ] Rep nullifier: Poseidon(4)
- [ ] Blinded user state: Poseidon(4)
- [ ] Blinded hash chain: Poseidon(4)
- [ ] Graffiti: Poseidon(1)

File: packages/core/cli/verifyReputationProof.ts [hunk 1]
```diff
@@ -100,9 +93,9 @@ const verifyReputationProof = async (args: any) => {
         return
     }
 
-    console.log(`Epoch key of the user: ${epk}`)
+    console.log(`Epoch key of the user: ${reputationProof.epochKey}`)
     console.log(
-        `Verify reputation proof from attester ${attesterId} with min rep ${minRep}, reputation nullifiers amount ${repNullifiersAmount} and graffiti pre-image ${args.graffiti_preimage}, succeed`
+        `Verify reputation proof from attester ${reputationProof.attesterId} with min rep ${reputationProof.minRep}, reputation nullifiers amount ${reputationProof.proveReputationAmount} and graffiti pre-image ${args.graffiti_preimage}, succeed`
```
Comment (vivianjeng) [inline packages/core/cli/verifyReputationProof.ts:98]: `args.graffiti_preimage` should be `reputationProof.graffitiPreImage`

File: packages/contracts/contracts/Unirep.sol [hunk 2]
```diff
@@ -386,15 +386,16 @@ contract Unirep is IUnirep, zkSNARKHelper, VerifySignature {
 
     /**
      * @dev A user spend reputation via an attester, the non-zero nullifiers will be processed as a negative attestation
-     * publicSignals[0: maxReputationBudget ] = [ reputationNullifiers ]
-     * publicSignals[maxReputationBudget    ] = [ epoch ]
-     * publicSignals[maxReputationBudget + 1] = [ epochKey ]
-     * publicSignals[maxReputationBudget + 2] = [ globalStateTree ]
+     * publicSignals[0] = [ epochKey ]
+     * publicSignals[1] = [ globalStateTree ]
```
Comment (vivianjeng) [inline packages/contracts/contracts/Unirep.sol:390]: I hope that the `=` and `[`, `]` can be aligned

File: packages/circuits/circuits/verifyEpochKey.circom [hunk 3]
```diff
@@ -1,36 +1,46 @@
 /*
     Verify if epoch key is computed correctly
-    epoch_key = hash5(id_nullifier, epoch, nonce);
+    epoch_key = hash2id_nullifier + nonce, epoch);
```
Comment (vivianjeng) [inline packages/circuits/circuits/verifyEpochKey.circom:3]: `hash2(id_nullifier + nonce, epoch)`

File: packages/circuits/test/utils.ts [hunk 4]
```diff
@@ -229,13 +224,13 @@ const genEpochKeyCircuitInput = (
     const circuitInputs = {
         GST_path_elements: proof.siblings,
         GST_path_index: proof.pathIndices,
-        GST_root: root,
+        // GST_root: root,
```
Comment (vivianjeng) [inline packages/circuits/test/utils.ts:227]: comments for debugging can be removed 👍🏻

**Review by vivianjeng — APPROVED** [review]
LGTM 
Good PR

**Pull request merged**